android/src/main/java/com/oblador/keychain/KeychainModule.java

```diff
@@ -905,6 +1018,69 @@ public void waitResult() {
 
       Log.i(KEYCHAIN_MODULE, "unblocking thread.");
     }
+
+    //region Deeply nested declarations for device workarounds
+
+    /**
+     * Class that handles authentication and decrypting when fingerprint authentication succeeds.
+     *
+     * Does nothing on its own and instead defers to given method references for implementation details, acting as an
+     * interface implementation for authentication purposes.
+     */
+    private class FingerprintManagerListener extends FingerprintManagerCompat.AuthenticationCallback {
+      @Override
+      public void onAuthenticationError(int errMsgId, CharSequence errString) {
+        final WritableMap params = KeychainModule.this.createOnFailureEventParams(true, errString);
+
+        KeychainModule.this.sendNativeEvent(EventNames.ON_FALLBACK_AUTHENTICATION_FAILURE, params);
+      }
+
+      @Override
+      public void onAuthenticationHelp(int helpCode, CharSequence helpString) {
+        final WritableMap params = KeychainModule.this.createOnFailureEventParams(false, helpString);
+
+        KeychainModule.this.sendNativeEvent(EventNames.ON_FALLBACK_AUTHENTICATION_FAILURE, params);
+      }
+
+      @Override
+      public void onAuthenticationSucceeded(FingerprintManagerCompat.AuthenticationResult result) {
+        InteractiveBiometric.this.handleAuthenticationSuccess();
+
+        if (InteractiveBiometric.this.getError() != null) {
+          final WritableMap params = KeychainModule.this.createOnFailureEventParams(false, null);
+
+          KeychainModule.this.sendNativeEvent(
+            EventNames.ON_FALLBACK_AUTHENTICATION_FAILURE,
+            params
+          );
+        } else if (InteractiveBiometric.this.getResult() != null) {
+          KeychainModule.this.sendNativeEvent(EventNames.ON_FALLBACK_AUTHENTICATION_SUCCESS,null);
+        }
+      }
+
+      @Override
+      public void onAuthenticationFailed() {
+        final WritableMap params = KeychainModule.this.createOnFailureEventParams(false, null);
+
+        KeychainModule.this.sendNativeEvent(EventNames.ON_FALLBACK_AUTHENTICATION_FAILURE, params);
+      }
+
+      public void authenticate() {
+        try {
+          KeychainModule.this.signal = new CancellationSignal();
+
+          final FingerprintManagerCompat fm = FingerprintManagerCompat.from(getReactApplicationContext());
+
+          KeychainModule.this.sendNativeEvent(EventNames.ON_FALLBACK_AUTHENTICATION_START, null);
```

Comment:
There's a timing issue here. By the time the event goes over the bridge, the main UI thread is already blocked waiting on authentication results. Chances are I need to move this further up, or fire it way before and wait for something, not sure.

Reply:
Took a look, I was on the right track with threading. This whole thing needs to be rewritten to not block on a single call and instead store the promise to be resolved later once authentication succeeds / fails.

As it is, it blocks everything on the native bridge until things pass. This is fine when using the `BiometricPrompt` since that displays a system UI which takes priority, but when trying to do so with the fingerprint manager, it blocks the RN thread making the render hang until a scan succeeds.

This is a pretty big lift: needs to consider fetching the item in keystore again and loading it on authentication success or failure. It also needs to be compatible with the `BiometricPrompt` flow which should block the thread to keep it simple.